firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/NewReleaseFetcher.java

```diff
@@ -62,31 +62,42 @@ public Task<AppDistributionReleaseInternal> checkForNewRelease() {
                 .fetchNewRelease()
                 .onSuccessTask(
                     lightweightExecutor,
-                    release -> Tasks.forResult(isNewerRelease(release) ? release : null)));
+                    release ->
+                        isNewerRelease(release)
+                            .onSuccessTask(
+                                lightweightExecutor,
+                                isNewer -> Tasks.forResult(isNewer ? release : null))));
   }
 
-  private boolean isNewerRelease(AppDistributionReleaseInternal retrievedNewRelease)
+  private Task<Boolean> isNewerRelease(AppDistributionReleaseInternal retrievedNewRelease)
       throws FirebaseAppDistributionException {
     if (retrievedNewRelease == null) {
       LogWrapper.v(TAG, "Tester does not have access to any releases");
-      return false;
+      return Tasks.forResult(false);
     }
 
     long newReleaseBuildVersion = parseBuildVersion(retrievedNewRelease.getBuildVersion());
 
     if (isOlderBuildVersion(newReleaseBuildVersion)) {
       LogWrapper.v(TAG, "New release has lower version code than current release");
-      return false;
+      return Tasks.forResult(false);
     }
 
     if (isNewerBuildVersion(newReleaseBuildVersion)
-        || !isSameAsInstalledRelease(retrievedNewRelease)
         || hasDifferentAppVersionName(retrievedNewRelease)) {
-      return true;
-    } else {
-      LogWrapper.v(TAG, "New release is older or is currently installed");
-      return false;
+      return Tasks.forResult(true);
     }
+
+    return isSameAsInstalledRelease(retrievedNewRelease)
+        .onSuccessTask(
+            lightweightExecutor,
+            isSame -> {
+              if (isSame) {
+                LogWrapper.v(TAG, "New release is older or is currently installed");
```

Comment:
Is "older" always correct? Could be "newer" as well, right?

Reply:
Ah sorry I meant to clarify this because I knew it was misleading. It can be "older" in that if we can return `false` here only if the release has the same app version name, and the same APK hash, and has either an **equal or older build version**. If the build version was newer it would have returned `true` in the condition above.

I've added a comment and renamed some methods to make it clearer what's going on here.